proof/invariant-abstract/ARM_HYP/ArchArch_AI.thy

```diff
@@ -983,24 +987,179 @@ lemma valid_global_vspace_mappings_vcpu_update_str:
+  for cap_refs_respects_device_region[wp]: cap_refs_respects_device_region
+  (wp: crunch_wps cap_refs_respects_device_region_dmo
+   simp: set_cntv_cval_64_def read_cntpct_def set_cntv_off_64_def
+         get_irq_state_def maskInterrupt_def)
+
+lemma arm_current_vcpu_update_valid_global_objs[simp]:
+  "valid_global_objs (s\<lparr>arch_state := arm_current_vcpu_update f (arch_state s)\<rparr>) = valid_global_objs s"
+  by (clarsimp simp: valid_global_objs_def)
+
+crunches associate_vcpu_tcb
+  for valid_global_objs[wp]: valid_global_objs
+  (wp: crunch_wps device_region_dmos)
+
+lemma set_vcpu_tcb_Some_hyp_live[wp]:
+  "\<lbrace>\<top>\<rbrace> set_vcpu vcpu (v\<lparr>vcpu_tcb := Some tcb\<rparr>) \<lbrace>\<lambda>_. obj_at hyp_live vcpu\<rbrace>"
+  apply (wpsimp wp: set_vcpu_wp)
+  apply (clarsimp simp: obj_at_def hyp_live_def arch_live_def)
+  done
+
+lemma associate_vcpu_tcb_valid_arch_state[wp]:
+  "associate_vcpu_tcb vcpu tcb \<lbrace>valid_arch_state\<rbrace>"
+  apply (clarsimp simp: associate_vcpu_tcb_def)
+  apply (wpsimp wp: vcpu_switch_valid_arch)
+        apply (rule_tac Q="\<lambda>_. valid_arch_state and obj_at hyp_live vcpu" in hoare_post_imp)
+         apply fastforce
+        apply wpsimp
+       apply (wpsimp wp: arch_thread_set.valid_arch_state)
+      apply (wpsimp wp: arch_thread_set_wp)+
+  done
+
+lemma dmo_valid_machine_state[wp]:
+  "do_machine_op (set_cntv_cval_64 w) \<lbrace>valid_machine_state\<rbrace>"
+  "do_machine_op read_cntpct \<lbrace>valid_machine_state\<rbrace>"
+  "do_machine_op (set_cntv_off_64 w') \<lbrace>valid_machine_state\<rbrace>"
+  "do_machine_op (maskInterrupt m irq) \<lbrace>valid_machine_state\<rbrace>"
+  "do_machine_op (setHCR word) \<lbrace>valid_machine_state\<rbrace>"
+  "do_machine_op isb \<lbrace>valid_machine_state\<rbrace>"
+  "do_machine_op dsb \<lbrace>valid_machine_state\<rbrace>"
+  "do_machine_op (set_gic_vcpu_ctrl_hcr f) \<lbrace>valid_machine_state\<rbrace>"
+  "do_machine_op (set_gic_vcpu_ctrl_lr n w'') \<lbrace>valid_machine_state\<rbrace>"
+  "do_machine_op (set_gic_vcpu_ctrl_apr w''') \<lbrace>valid_machine_state\<rbrace>"
+  "do_machine_op (set_gic_vcpu_ctrl_vmcr w''') \<lbrace>valid_machine_state\<rbrace>"
+  "do_machine_op (get_gic_vcpu_ctrl_lr w'''') \<lbrace>valid_machine_state\<rbrace>"
+  "do_machine_op get_gic_vcpu_ctrl_apr \<lbrace>valid_machine_state\<rbrace>"
+  "do_machine_op get_gic_vcpu_ctrl_vmcr \<lbrace>valid_machine_state\<rbrace>"
+  "do_machine_op get_gic_vcpu_ctrl_hcr \<lbrace>valid_machine_state\<rbrace>"
+  unfolding valid_machine_state_def set_cntv_cval_64_def read_cntpct_def set_cntv_off_64_def
+            maskInterrupt_def setHCR_def set_gic_vcpu_ctrl_hcr_def set_gic_vcpu_ctrl_lr_def
+            set_gic_vcpu_ctrl_apr_def set_gic_vcpu_ctrl_vmcr_def get_gic_vcpu_ctrl_lr_def
+            get_gic_vcpu_ctrl_apr_def get_gic_vcpu_ctrl_vmcr_def get_gic_vcpu_ctrl_hcr_def
+  by (wpsimp wp: hoare_vcg_all_lift hoare_vcg_disj_lift dmo_machine_state_lift)+
```

Comment:
@corlewis : do you have any thoughts on crunch and doing this kind of thing? it seems rather repetitive and the `w''''` is indicative of that

Reply:
Unfortunately, `crunch` currently wouldn't be any help here unless it turned out that you could prove `do_machine_op f ⦃valid_machine_state⦄`. There isn't a way to specify parameters for the functions being crunched.

Reply:
Let me know if I should experiment with another way of doing this